### ISSUE TITLE
Spring formatter not called in alle cases anymore after upgrade to thymeleaf 3

### ISSUE THREAD

**Issue by SnakeSVx** [thymeleaf/thymeleaf-spring#128]
After switching to thymeleaf 3 we have some cases where the formatter is not called anymore when using *{{expression}}.

Before on thymeleaf 2 the same code worked without any problems.

While trying to find out where the problem was coming from, I noticed that the "BindStatus.editor" field is null on line 165 of org.thymeleaf.spring4.expression.SPELVariableExpressionEvaluator.

As the editor is null, ValueFormatterWrapper.getDisplayString return with just a toString of the field instead of calling the Spring formatter for this.

Edit:
From what I can notice, it seems like in thymeleaf 2, the expression was using the actual class of the property to find the "formatter" to use, while now it is trying to use the class in the field definition.

In our case the field has a type that is an abstract superclass which has no formatter.

Could the old thymeleaf 2 behaviour be brought back where it uses the actual class to get the formatter instead of the class defined on the field trying to be used?

**Comment by danielfernandez** [comment]
Could you please confirm what specific version of Thymeleaf 3 you are using? is it `3.0.3.RELEASE`? 

Given your expression is a selection expression (`*{{...}}`), and therefore your property is nested, you might be suffering from #112, fixed in Thymeleaf 2.1.5 and 3.0.1.

Also, the line you specify actually determines if the property in the expression is relative to a form-bound object... is your property bound to a form object you are selecting previously with `th:object` on a `<form>`?

**Comment by SnakeSVx** [comment]
We are using Thymeleaf 3, version 3.0.3.RELEASE.
As usage we have 2 cases that seems to fail using the same object.
1. Without form, but using th:object: <div th:object="${object}><span th:text="*{{property}}">default</span> .... </div>
2. With a form and th:object < form th:object="${object}">< p class="readonly_field" th:text="*{{property}}">readonly< /p> .... < /form>

The property itself is defined as "private A property", the actual value is class B (which extends A).
A has no formatter available, B has a formatter.

Other environment factors:
We are using spring-boot 1.4.3 (spring 4.3.5).

**Comment by SnakeSVx** [comment]
More test results:
Without using th:object and the *{{ }} notation it seems to still work without a problem.
So doing `<p th:text="${{object.property}}">...</p>`, the correct formatter is called, in that case it skips the entire block the code on line 165 is in and continues to line 277
 ```
if (conversionService instanceof SpringStandardConversionService) {
                // The conversion service is a mere bridge with the Spring ConversionService, therefore
                // this makes use of the complete Spring type conversion infrastructure, without needing
                // to manually execute the conversion.
                return exp.expression.getValue(thymeleafEvaluationContext, evaluationRoot, String.class);
            }
```

In Thymeleaf 2.1.5:

It never reaches SpelVariableExpressionEvaluator for evaluating the expression.

**Comment by danielfernandez** [comment]
I'm trying to verify this mirroring your scenario in which the *containing* bean has a property of a superclass, the objects in that property are of a subclass, and the `Formatter` is applied to the subclass only… 

But right now I'm seeing exactly the same behaviour in Thymeleaf 2.1.5 and 3.0.3. Only when the field in the containing bean is declared to be of the subclass (the exact one that has the `Formatter`), conversions are performed. And it works the same with `th:object` and `*{{...}}`, without `th:object` and `${{bean....}}` and inside a form field.

Actually, this type of conversion should be delegated to Spring's EL engine itself, so it's Spring who works that way. And I don't think this has changed since 2.1... but anyway, obviously there is something different happening to you between 2.1.5 and 3.0.3 (assuming also that you have not changed the Spring version), so I'll have to investigate a bit more and see what can be happening…

**Comment by danielfernandez** [comment]
I'm still unable to replicate the difference in behaviour you report, I'm afraid.

I've just created a test repository that should allow us better determine the cause of what you are experiencing. Please find it here: https://github.com/danielfernandez/test-nested-springel-conversion

As explained in the README, it contains two applications (actually the same one, but with Thymeleaf 2.1.5 and 3.0.3 respectively), simulating what appears to be your scenario: a `Field` class extending `AbstractField`, and a `Bean` class which contains two fields: one of type `Field` and another one of type `AbstractField`. There is a `Formatter<Field>` formatter, which applies only to the `Field` field in the bean. And it shows the same results in 2.1.5 and 3.0.3.

Please feel free to modify the application to your needs in order to replicate the issue, and add a PR if you can so that I can better understand what is going on…

**Comment by danielfernandez** [comment]
Note tests referenced above were wrongly linked to this ticket. They were actually for thymeleaf/thymeleaf#570

**Comment by SnakeSVx** [comment]
I'll see what I can do to reproduce it using the testcases above.
I'll also start checking some of our other configuration to see if anything changed in any other area that might be related to this issue.

**Comment by SnakeSVx** [comment]
Haven't been able to reproduce the current situation we have with thymeleaf 2 yet, so it must be something we have that is configuring something that works in thymeleaf 2 but not thymeleaf 3.

On the other hand, I have found a small workaround, that "fixes" the issue on both thymeleaf 2 and thymeleaf 3 in our case. One thing I noticed as a different is that we are using fragments to generate the "readonly" div, inside of this fragment we are using a preprocessor for the "field" name.

I commited an example on my fork of the test repository:
https://github.com/SnakeSVx/test-nested-springel-conversion

The weird thing about it is that by default the fragment works the same as when not using the fragement, but once we add an empty string in the mix, suddenly it behaves like how we are seeing it in our current production environment with thymeleaf 2.

So in short:
`<span th:text="*{{__${field}__}}">div</span>` does not format as expected
`<span th:text="*{{''+__${field}__}}">div</span>` formats as expected

**Comment by danielfernandez** [comment]
Any news on this?

As for why `<span th:text="*{{''+__${field}__}}">` does indeed apply the converter even when the field is abstract (which is what I observe), this is due to the fact that `'' + fieldAbs` does not have a registered `BindStatus` (it's not a valid field name), whereas `fieldAbs` does have one because it is a valid field name. In the case where no `BindStatus` is registered the type metadata of the bean cannot be used, obviously, so the type is inferred from the object itself (thus it doesn't matter that the bean field is typed to the abstract class).

So IMHO this makes sense. Also, it works exactly the same in 2.1 and 3.0.

**Comment by danielfernandez** [comment]
Closing this for now as *non reproducible*. Please don't hesitate to reopen it in the future if you are able to reproduce it.

**Comment by cassimirojackson** [comment]
I had a similar issue and I solved using some parts of renderFragment method from ThymeleafView:

```
WebContext ctx = new WebContext(request, response, request.getServletContext());
		
		final ConversionService conversionService =
                (ConversionService) request.getAttribute(ConversionService.class.getName()); // might be null!
        final ThymeleafEvaluationContext evaluationContext =
                new ThymeleafEvaluationContext(applicationContext, conversionService);
        ctx.setVariable(ThymeleafEvaluationContext.THYMELEAF_EVALUATION_CONTEXT_CONTEXT_VARIABLE_NAME, evaluationContext);
```
This  way, when  my context has a  instance of ThymeleafEvaluationContext all my formatters works fine.

If anyone is still interested in this issue I can provide more information.
